packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m

```diff
@@ -143,96 +163,78 @@ - (NSDictionary *)constantsToExport {
-      } else {
-        authOptions = UNAuthorizationOptionAlert | UNAuthorizationOptionSound | UNAuthorizationOptionBadge;
+        options |= UNAuthorizationOptionProvisional;
       }
+    }
 
-      [[UNUserNotificationCenter currentNotificationCenter] requestAuthorizationWithOptions:authOptions completionHandler:^(BOOL granted, NSError *_Nullable error) {
-        if (error) {
-          [RNFBSharedUtils rejectPromiseWithNSError:reject error:error];
-        } else {
-          [[RNFBMessagingAppDelegateInterceptor sharedInstance] setPromiseResolve:customResolver andPromiseReject:reject];
-          dispatch_async(dispatch_get_main_queue(), ^{
-            [[UIApplication sharedApplication] registerForRemoteNotifications];
-          });
-        }
-      }];
+    if ([permissions[@"announcement"] isEqual:@(YES)]) {
+      if (@available(iOS 13.0, *)) {
+        #if __IPHONE_OS_VERSION_MAX_ALLOWED >= 130000
+        options |= UNAuthorizationOptionAnnouncement;
+        #endif
+      }
+    }
 
-    } else {
-      // TODO community iOS 9 support could be added here
-      [RNFBSharedUtils rejectPromiseWithUserInfo:reject userInfo:[@{
-          @"code": @"unsupported-platform-version",
-          @"message": @"requestPermission call failed; minimum supported version requirement not met (iOS 10)."} mutableCopy]];
+    if ([permissions[@"carPlay"] isEqual:@(YES)]) {
+      options |= UNAuthorizationOptionCarPlay;
     }
-    
+
+    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
+    [center requestAuthorizationWithOptions:options completionHandler:^(BOOL granted, NSError *_Nullable error) {
+      if (error) {
+        [RNFBSharedUtils rejectPromiseWithNSError:reject error:error];
+      } else {
+        [self hasPermission:resolve :reject];
+      }
+    }];
+  } else {
+    [RNFBSharedUtils rejectPromiseWithUserInfo:reject userInfo:[@{
+        @"code": @"unsupported-platform-version",
+        @"message": @"requestPermission call failed; minimum supported version requirement not met (iOS 10)."} mutableCopy]];
   }
 }
 
 RCT_EXPORT_METHOD(registerForRemoteNotifications:
   (RCTPromiseResolveBlock) resolve
     : (RCTPromiseRejectBlock) reject
 ) {
-  [[RNFBMessagingAppDelegateInterceptor sharedInstance] setPromiseResolve:resolve andPromiseReject:reject];
-  [[UIApplication sharedApplication] registerForRemoteNotifications];
-  
-  if ([UIApplication sharedApplication].isRegisteredForRemoteNotifications == YES) {
-    return resolve(@([RCTConvert BOOL:@(YES)]));
-  }
+  #if TARGET_IPHONE_SIMULATOR
+  resolve(@([RCTConvert BOOL:@(YES)]));
+  return;
+  #endif
   if (@available(iOS 10.0, *)) {
-    UNAuthorizationOptions authOptions;
-    if (@available(iOS 12.0, *)) {
-      authOptions = UNAuthorizationOptionProvisional | UNAuthorizationOptionAlert | UNAuthorizationOptionSound | UNAuthorizationOptionBadge;
+    if ([UIApplication sharedApplication].isRegisteredForRemoteNotifications == YES) {
```

Comment:
This condition is broken and is the _real_ root cause of the 1001 error code.
Diving into the code of the iOS SDK reveals the 1001 code is [kFIRInstanceIDErrorCodeMissingAPNSToken](https://github.com/firebase/firebase-ios-sdk/blob/42c5ae17fc71886b23ba008623f509db0ae5316a/Firebase/InstanceID/NSError%2BFIRInstanceID.h#L43), which is thrown by the `tokenWithAuthorizedEntity` method if it did not receive an APNS token, which is provided by the `didRegisterForRemoteNotificationsWithDeviceToken` callback: https://github.com/firebase/firebase-ios-sdk/blob/c049f741f0d3b7de955dff7c4173baedf3aa8355/Firebase/InstanceID/FIRInstanceID.m#L296

While `instanceIDWithHandler` has a caching system that will auto-fill the APNS token, which is why #3130 worked, we should **not** bypass registering for notifications in any case since iOS will reply with an updated token if needed.
According to the [official docs](https://developer.apple.com/documentation/usernotifications/registering_your_app_with_apns#2942135):
> You register your app and receive your device token **each time your app launches** using Apple-provided APIs.